### PR TITLE
[okina] Umpire memory manager for okina [jdahm:umpire-mm-dev]

### PR DESCRIPTION
**Pull request by jdahm**
This adds Umpire support to MFEM's memory manager in the okina branch.

### CODE REVIEW

File: config/defaults.mk [hunk 1]
```diff
@@ -291,6 +292,10 @@ ifeq ($(MFEM_USE_OCCA),YES)
   OCCA_LIB := -Wl,-rpath,$(OCCA_DIR)/lib -L$(OCCA_DIR)/lib -locca
 endif
 
+UMPIRE_DIR = @MFEM_DIR@/../umpire
+UMPIRE_OPT = -I$(UMPIRE_DIR)/include
+UMPIRE_LIB = -L$(UMPIRE_DIR)/lib -lumpire -lumpire_op -lumpire_resource -lumpire_strategy -lumpire_tpl_judy -lumpire_util
+
```
Comment (tzanio) [inline config/defaults.mk]: Are all of this libraries really necessary... we can't just link with `-lumpire`?

Comment (jdahm) [reply config/defaults.mk]: Thanks for taking a look! Currently these are all necessary. That's slated to change very soon in Umpire. After that you're correct, all we'll need is `-lumpire`. Stay tuned!

**Comment by jdahm** [comment]
This is ready to be looked at concerning the overall style of the memory manager abstraction.

I'm certain some of the logic, especially in the `getOccaPtr()` method, is incomplete and I didn't implement the async transfer yet. I've been testing this locally, so this still needs to verified with GPUs on Lassen before merging.

**Comment by artv3** [comment]
@jdahm thanks for putting this together so quickly! To try this out would I just link in Umpire and run example 1 with the GPU enabled? I'm looking forward to trying this in my application.

**Comment by jdahm** [comment]
@camierjs Wow, you did a ton of work on this today! Thank you for all the help toward this integration. I'll also test as soon as possible.

**Comment by jdahm** [comment]
@artv3 Yes, enabling umpire in the MFEM build automatically enables it as the memory manager type (i.e. it's a typedef in the code). Therefore, to enable, just build and install umpire using `-D ENABLE_CUDA=ON` (`develop` branch), then configure the `umpire-mm-dev` branch with `MFEM_USE_MM=YES MFEM_USE_UMPIRE=YES UMPIRE_DIR=${path_to_umpire_install}`.

**Comment by jdahm** [comment]
I finally updated this branch to the latest version of okina, but there have been significant changes to the memory manager there that did not build on the clean-up done in this branch. This seems to have introduced a regression - example 1 now fails to work correctly.

**Comment by jdahm** [comment]
Even if the code enabled by setting `MFEM_USE_UMPIRE=YES` are not entirely working, I think we should merge this into `okina` soon, since it documents the memory manager and puts a layer of separation between the interface and implementation, which allows for potentially different future versions. We can then continue clean-up without being side-tracked by merge conflicts.

**Comment by artv3** [comment]
Do we have a sense of how far we are from getting ex 1 to work? I am for merging this soon, but I'm a bit hesitant on merging something that doesn't quite work.

**Comment by tzanio** [comment]
Outdated due to new memory handling in mfem-4.0.

**Pull request closed without merging**